posts/Stef/2019-11-18-hibernate-orm-with-panache-in-quarkus.adoc

```diff
@@ -0,0 +1,465 @@
+        return find("lastName", lastName).list();
+    }
+    
+    public List<Person> findBornAfter(Date date){
+        return find("birth > :date", Parameters.with("birth", date)).list();
+    }
+}
+----
+
+Note that the `find` convenience method allows HQL, but also _simplified_ HQL (think of it as contextualized HQL):
+
+- If your query is empty, it expands to `FROM <entityType>`
+- If your query starts with `FROM` or `SELECT`, it is left alone as HQL
+- If your query starts with `ORDER BY...` it expands to `FROM <entityType> ORDER BY...`
+- If your query only has a single property and argument, it expands to `FROM <entityType> WHERE <property> = <argument>`
+- Otherwise, your query is taken as a `WHERE...` clause and expands to `FROM <entityType> WHERE...`
+
+This allows many simple queries to be simplified to a minimum, while allowing complex queries to be left as-is.
+
+Now, our REST endpoint is not changed much, but let's include it for good measure:
+
+[source%nowrap,java]
+----
+@Path("/")
+public class PersonEndpoint {
+    @Inject
+    private PersonDao personDao;
+    
+    @GET
+    @Path("people")
+    public List<Person> all(){
+        return personDao.findAll().list();
+    }
+
+    @GET
+    @Path("people/by-name")
+    public List<Person> findByName(@PathParam String name){
+        return personDao.findByName(name);
+    }
+
+    @GET
+    @Path("people/born-after")
+    public List<Person> findBornAfter(@PathParam Date date){
+        return personDao.findBornAfter(date);
+    }
+
+    @GET
+    @Path("person/{id}")
+    public Person findById(@PathParam Long id){
+        Person p = personDao.findById(id);
+        if(p == null)
+            throw new WebApplicationException(Status.NOT_FOUND);
+        return p;
+    }
+
+    @PUT
+    @Path("person/{id}")
+    public void updatePerson(@PathParam Long id, Person newPerson){
+        Person p = personDao.findById(id);
+        if(p == null)
+            throw new WebApplicationException(Status.NOT_FOUND);
+        p.birth = newPerson.birth;
+        p.firstName = newPerson.firstName;
+        p.lastName = newPerson.lastName;
+    }
+
+    @DELETE
+    @Path("person/{id}")
+    public void deletePerson(@PathParam Long id){
+        Person p = personDao.findById(id);
+        if(p == null)
+            throw new WebApplicationException(Status.NOT_FOUND);
+        personDao.delete(p);
+    }
+    
+    @POST
+    @Path("people")
+    public Response newPerson(@Context UriInfo uriInfo, Person newPerson){
+        Person p = new Person();
+        p.birth = newPerson.birth;
+        p.firstName = newPerson.firstName;
+        p.lastName = newPerson.lastName;
+        personDao.persist(p);
+        
+        URI uri = uriInfo.getAbsolutePathBuilder()
+                .path(PersonEndpoint.class)
+                .path(PersonEndpoint.class, "findById")
+                .build(p.id);
+        return Response.created(uri).build();
+    }
+}
+----
```

Comment:
so - personal opinion here.

We're trying hard to educate people that exposing entities directly to the REST layer is a terrible idea, not leat because it actualy is a terrible idea to couple the API to the RDBMS schema, but also as there's various unsolved challenges with Jackson integration and what to do with transaction boundaries spanning beyond lazy loading..

could we avoid exposing it like this? Maybe the same example still makes sense even without the JAX annotaitons?

Reply:
OK, I did wonder about this, and I thought I would add a disclaimer to not do it this way, but didn't want to add DTOs in the example because they'd muddle the blog further. I forgot to do it, but let's add it and tell me if you think it's enough.

Reply:
Add what? 

I'd agree to not add DTOs as that would add a fair bit of noise, but why simply not exposing a rest endpoint? You could focus on showing the DAO code as a hipotethical useful bean, rather than an anemic object.

Reply:
I've added a note about DTOs. I thought about not showing the REST endpoint, but:

- I wanted to show that the generated methods covered all CRUD REST endpoints use-cases
- I wanted to show usage (inject DAO, use it, then replace with static methods. Same for getter/setter)

I could make the endpoint smaller but it'd undermine the first point. I could avoid repeating methods that are unchanged when I show the endpoint again later, but then users would not be able to copy/paste that class if they want to.

So, not really sure if I should remove the rest endpoints or not. I'd lean towards keeping them, but not strongly.

Reply:
Ok, as you prefer; it was just a thought - not a strong one either.